### PR TITLE
Add binding for `setSerialMessageCallback`

### PR DESCRIPTION
**Pull request by paulyoung**
Adds bindings and an example for `playdate->system->setSerialMessageCallback`

Please also see https://devforum.play.date/t/feature-request-add-a-userdata-argument-to-setserialmessagecallback/17333

Thanks!

### CODE REVIEW

File: api/system/src/lib.rs [hunk 1]
```diff
@@ -247,10 +248,25 @@ impl<Api: api::Api> System<Api> {
...
+	//
+	//   pub fn set_serial_message_callback<Fn>(&self, callback: Fn) where Fn: FnMut(String) {
+	//
+	// but we can't do that because of
+	// https://devforum.play.date/t/feature-request-add-a-userdata-argument-to-setserialmessagecallback/17333
+	pub fn set_serial_message_callback(&self, callback: Option<unsafe extern "C" fn(data: *const c_char)>) {
+		let f = self.0.set_serial_message_callback();
+		unsafe { f(callback) }
+	}
+}
```
Comment (boozook) [inline api/system/src/lib.rs:266]: I agree that having userdata would be really helpful. But it is __not critically__ necessary. You can do without it, albeit with a little overhead.

We still can use [typed storage][storage], that's compiletime kinda magic. If explain shortly, any `Fn` type in the two levels of AST in the rustc have unique type, so we can 
- store it with unique key
- store type as type-param (generic)
- get it back with type-param 

\* _i mainly mean closures, not fn-pointers_

For example see my implementation in scoreboards - [storage][] and [usage][].

_This comment — it's just a comment and a suggestion for discussion, not an urgent demand for change, of course._ 😬

[storage]: https://github.com/boozook/playdate/blob/main/api/scoreboards/src/storage.rs
[usage]: https://github.com/boozook/playdate/blob/main/api/scoreboards/src/lib.rs#L83-L85

Comment (boozook) [reply api/system/src/lib.rs:266]: Also, I don't see any reason for Dave to add userdata for that callback. On one side, hi stores just one ptr, but on other side hi have to store ptr and mapping like ptr->userdata. I'm sure he'll say it's an overhead and won't do it without more reasons. 🤷🏻‍♂️

**Comment by boozook** [comment]
#328 

Configuration env vars: 
- `PD_MOUNT_TIMEOUT`, default `1000` ms
- `PD_UNMOUNT_TIMEOUT`, default `1000` ms.

Values in __milliseconds__.

File: api/system/src/lib.rs (hunk 1 above)
Comment (paulyoung) [reply api/system/src/lib.rs]: Thanks! Please see c6ec4474a39c4ddd443abf1b37d4181f7d07f69f

**Comment by paulyoung** [comment]
Relates to #283

**Comment by boozook** [comment]
Thank you so much! I have to play with it and think about it a little bit.

**Review by boozook — COMMENTED** [review]
I reviewed your solution as well as entire task and have to say that I was wrong when suggested "typed storage".
There is should be solution dramatically simpler.
Also in my example all callbacks are "oneshot" and removing immediately after call with `store.remove::<F>()` and later entire storage with `clean_store`. That's not needed for this case. This is like update callback registering, so we should be able set callback, receive calls multiple times and optionally "unsubscribe".


I've test it this way:
```bash
# Build, install, run:
cargo playdate run -p=playdate-system --example=set-serial-message-callback --features=sys/lang-items,sys/entry-point --device
--no-read

# Then send messages:
pdtool send msg "one" && pdtool read
pdtool send msg "two" && pdtool read
pdtool send msg "two" && pdtool read
# ...

# Output:
# serial_message_callback: one
# PANIC: [api/system/src/storage.rs@24] "missed callback"
```

- - -

I've created [PR](https://github.com/paulyoung/playdate/pull/1) for your PR ;)

**Comment by paulyoung** [comment]
Strange, I'm running into #326 on this branch despite having pulled in the fixes for that issue.

**Comment by paulyoung** [comment]
For some reason I'm still hitting `ERROR: missing field `serial_num` at line 31 column 9`

**Comment by paulyoung** [comment]
@boozook thanks for the PR. I was thinking this should be similar to `set_update_callback` but have been sick for a few days and didn't follow up.

**Comment by paulyoung** [comment]
To clarify, when I run `cargo playdate run -p=playdate-system --example=set-serial-message-callback --features=sys/lang-items,sys/entry-point --device --no-read` I hit the `ERROR: missing field `serial_num` at line 31 column 9` error but otherwise everything on this branch seems to work for me so I don't think that's related or should hold up merging this PR.

I can try again once it's on `main` and follow up in #326.

**Comment by boozook** [comment]
I hope you doing well. 

Just to be sure:
Did you install `cargo-playdate` from this branch? If not so, plz do `cargo run -p=cargo-playdate -- run -p=playdate-system --example=set-serial-message-callback --features=sys/lang-items,sys/entry-point --device --no-read`.

**Review by boozook — APPROVED** [review]

**Comment by paulyoung** [comment]
> Did you install `cargo-playdate` from this branch? 

@boozook I did. I'll try again now from `main` and report back in #326

**Pull request merged**